### PR TITLE
Implement COUPDAYS

### PR DESCRIPTION
**Pull request by natanhp**
## Description

Implement COUPDAYS function.

## Motivation and Context

This change will enable user to use COUPDAYS function from this library.

## How Has This Been Tested?

I tested it using the unit test, using screnario from [Microsoft Office COUPDAYS Official Documentation](https://support.microsoft.com/en-us/office/coupdays-function-cc64380b-315b-4e7b-950c-b30b0a76f671) and I cross checked it with Libre Office Calculator.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by nicolashefti** [comment]
Hi @natanhp, thanks for your contribution. I'm on vacations and will check that next week.

File: src/financial.js [hunk 1]
```diff
@@ -156,8 +189,47 @@ export function COUPDAYBS() {
  * @param {*} basis Optional. The type of day count basis to use.
  * @returns
  */
-export function COUPDAYS() {
-  throw new Error('COUPDAYS is not implemented')
+export function COUPDAYS(settlement, maturity, frequency, basis) {
+  basis = validateBasis(basis)
+
+  if (basis === 1) {
+    settlement = utils.parseDate(settlement)
```
Comment (nicolashefti) [inline src/financial.js:196]: Parameter checks (line 196 >> 206) could run for all basis values, no?

File: src/financial.js [hunk 2]
```diff
@@ -156,8 +189,47 @@ export function COUPDAYBS() {
...
+
+  let numOfDays
+
+  switch (basis) {
+    case 0:
+    case 2:
+    case 4:
+      numOfDays = 360
+      break
+    case 1:
```
Comment (nicolashefti) [inline src/financial.js:224]: Case 1 won't reach this line because of the if statement line 195 with return

File: src/financial.js [hunk 3]
```diff
@@ -156,8 +189,47 @@ export function COUPDAYBS() {
...
+    case 3:
```
Comment (nicolashefti) [inline src/financial.js:225]: Case 3 same as case 1 ? In this case why would it be different than the if statement above?

File: src/financial.js [hunk 4]
```diff
@@ -10,6 +10,45 @@ function ensureDate(d) {
   return d instanceof Date ? d : new Date(d)
 }
 
+// Calculate last coupon date before settlement
+function lastCoupDateBeforeSettlement(settlement, maturity, frequency) {
```
Comment (nicolashefti) [inline src/financial.js:14]: This is the core part of the COUPDAYS function. It won't be reused anywhere, so I would include this part in the function itself.

**Review by nicolashefti — CHANGES_REQUESTED** [review]
Thanks for your contribution. I'm not an Excel/financial functions expert, but I feel there is something wrong between the if and the switch statements. Let me know if I missed something.

File: src/financial.js (hunk 4 above)
Comment (natanhp) [reply src/financial.js:14]: Actually it can be used for any other functions such as COUPPCD

File: src/financial.js (hunk 3 above)
Comment (natanhp) [reply src/financial.js:224]: My bad, I should remove case 1 because it's already handled by the if statement in the line 195

**Review by nicolashefti — APPROVED** [review]
Thanks for the changes. It's good for me.

**Pull request merged**